introduce goto_programt::instructiont::type() and _nonconst()

kroening opened the pull request:
This introduces two methods to read and write the type of a GOTO program
instruction, and replaces various direct accesses.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
introduce goto_programt::instructiont::type() and _nonconst()

This introduces two methods to read and write the type of a GOTO program
instruction, and replaces various direct accesses.

peterschrammel commented on src/goto-instrument/uninitialized.cpp:
```diff
@@ -120,11 +120,11 @@ void uninitializedt::add_assertions(
         i1->code_nonconst() = code_declt(symbol_expr);
 
-        i2->type=ASSIGN;
+        i2->type_nonconst() = ASSIGN;
```
use the 2-param variant of `insert_after`?

peterschrammel commented on src/goto-instrument/uninitialized.cpp:
```diff
@@ -120,11 +120,11 @@ void uninitializedt::add_assertions(
 
         symbol_exprt symbol_expr(new_identifier, bool_typet());
-        i1->type=DECL;
+        i1->type_nonconst() = DECL;
```
use the 2-param variant of `insert_after`?

peterschrammel commented on jbmc/src/java_bytecode/remove_exceptions.cpp:
```diff
@@ -421,7 +421,7 @@ bool remove_exceptionst::instrument_throw(
     typecast_exprt(exc_expr, exc_thrown.type()));
   // now turn the `throw' into `assignment'
-  instr_it->type=ASSIGN;
+  instr_it->type_nonconst() = ASSIGN;
```
swap?

peterschrammel commented on src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp:
```diff
@@ -846,7 +846,7 @@ void disjunctive_polynomial_accelerationt::build_fixed()
     if(instruction.is_assert())
     {
-      instruction.type = ASSUME;
+      instruction.type_nonconst() = ASSUME;
```
turn_into_assume?

peterschrammel reviewed (commented):
💡 I think by introducing `turn_into_assume` (we've already got `turn_into_skip`), we can do without `type_nonconst()`.

kroening committed:
introduce instructiont::turn_into_assume

This method turns a GOTO or an ASSERT instruction into an ASSUME with the
same condition.  This is a common pattern in the code base, and introducing
the method enables doing so in a way that is guaranteed to preserve the
class invariants.

martin-cs commented on src/goto-programs/goto_program.h:
```diff
@@ -340,8 +340,24 @@ class goto_programt
+    /// Set the kind of the instruction.
+    /// This method is best avoided to prevent mal-formed instructions.
+    /// Consider using the goto_programt::make_X methods instead.
+    goto_program_instruction_typet &type_nonconst()
```
As this only has 7 users, would it be better to just refactor those and not have this functionality at all?  Alternatively, could this be marked as deprecated?

tautschnig replied:
Aren't we even down to 2 users, one of which can easily be fixed and the other being `read_bin_goto_object`? I think the latter could be handled by some low-level constructor that likely wants to be `private` with access only permitted via a `friend` declaration.

martin-cs reviewed (approved):
Seems reasonable.

kroening commented:
I've added `turn_into_assume`, which removes some usage of `type_nonconst()`.

tautschnig commented on unit/goto-programs/structured_trace_util.cpp:
```diff
@@ -56,7 +56,7 @@ TEST_CASE("structured_trace_util", "[core][util][trace]")
   add_instruction(loop_head_location, instructions);
   // 2: goto 1 # back_edge
   const auto back_edge = add_instruction(back_edge_location, instructions);
-  back_edge->type = GOTO;
+  back_edge->type_nonconst() = GOTO;
```
Can't we use `make_goto` for this one?

kroening replied:
add_instruction() in that unit test should be replaced altogether, but this goes beyond this PR.

tautschnig reviewed (commented):
It would be great to figure out a story for the `nonconst` case right away.

kroening commented:
Changing `read_bin_goto_object` is a major rejfactoring of that piece of code; much more involved than the initial scope of the PR. I'd do this separately.
